### ISSUE TITLE
Pipeline for patch-based machine learning

### ISSUE THREAD

**Issue by benjamindeleener** [spinalcordtoolbox/spinalcordtoolbox#1011]
This Patch-Based Machine Learning Pipeline will be integrated to SCT.
The pipeline is composed of several classes:
- FileManager: manages files in dataset, select patches and computes analytics on dataset.
- ClassificationModel: a template class for a machine learning model. A model must inherits from this class and implements two functions: `train(x, y)` and `y = predict(x)`
- Trainer: train the Model based on the dataset provided by the FileManager. Also performs hyperparameters optimization.
## FileManager

Inputs:
- Path to the dataset
- Path where to save results
- Function to explore the dataset. Input: path to the dataset. Output: list of dictionary with predetermined labels (ex: image, manual_seg, etc.)
- Extraction parameters. Examples: ratio training/testing/validation datasets, patches size, dimensions, orientations. Probably a dictionary. Patches are extrated using physical interpolation.
- Function to classify a patch based on ground truth. Input: patch + ground truth. Output: class
- List of processes to apply to each image. Examples: intensity normalization modeling (extract intensity histogram)

Outputs in text files:
- datasets.json
  List of images + ground truth, separated into training and testing dataset

``` python
{
'training': {'raw_images': [filename1, filename2, ...], 'gold_images': [filename1, filename2, ...]}, # filename1 could be replaced by a list if multimodal or multilabel
'testing': {'raw_images': [filename1, filename2, ...], 'gold_images': [filename1, filename2, ...]},
'dataset_path': path
}
```
- patches.json
  List of patches coordinates for each image + classes
  Dataset statistics. Examples: classes weights in each dataset.

``` python
{
'training': {'index in datasets.json:training': [coord[float, float, float], label(int); ...], ... }  # for each raw subject, the patch will be extracted in all raw and gold images, using physical coordinates.
'testing': {'index in datasets.json:testing': [coord[float, float, float], label(int); ...], ... }
'patch_info': {'patch_size': [int, int], 'patch_pixdim': {'orient1': [float, float], 'orient2': [float, float], 'orient3': [float, float]}}
'statistics': {'classes_training': {'class': [nb, weight]}, 'classes_testing': {'class': [nb, weight]}}
}
```
## ClassificationModel

Examples: Neural Network, SVM, Logistic Regression Model, etc.
Inputs:
- Model parameters. Examples: learning rate, number of iterations, class weights, tolerance

Must implements:
- Function `train(x, y)`: train the model on a dataset (x, y) where x is the patches and y is the labels (class)
- Function `y = predict(x)`: predict the label of a patch or a series of patches
## Trainer

The trainer will train the model on the training dataset with different sets of hyper-parameters and find the best solution with respect to the testing dataset.
Inputs:
- Path to dataset
- Dataset statistics computed by the FileManager
- Training and testing datasets computed by the FileManager
- Classification Model. Its hyper-parameters must be accessible.
- Normalization Model. This object implements a function `normalize()` that apply a intensity normalization model on a full 3D image. Example: intensity normalization model based on histogram-matching
- Feature extraction model. Takes a patch (or a series of patches) as input and outputs its (their) features. For convolutional networks, this model will be identity, i.e., the output will equal the input.
- Training parameters. Dictionary. Examples: `{'batch_size': 32, 'number_of_epochs': 100, 'number_of_iter_hyperparam_search':20}`
- Hyper-parameters ranges. Dictionary with the values of hyper-parameters to explore/optimize. Examples: `{'learning_rate':range(0.01, 1.0, 0.05), 'number_of_layers': range(2, 10, 1), 'number_of_features': [8, 16, 32, 64]}`

Outputs:
- Path to the best trained model along with its characteristics.
# To be continued...

**Comment by benjamindeleener** [comment]
We have an issue when saving the coordinates produced by the FileManager: as there is potentially millions of patches in one single image, the file `patches.json` is way too big.

The solution is to couple `pickle` and `bz2`:
```python
with bz2.BZ2File(path_output + 'patches.pbz2', 'w') as f:
    pickle.dump(global_results_patches, f)
```

**Comment by charleygros** [comment]
Prototype fonction Trainer.hyperparam_optimization:

hyperparam_dict: range dictionnary (for tuning parameters) or list of parameters (for only train)

```python
def hyperparam_optimization(self, fct_prepare_batch, nb_epoch, hyperparam_dict, ratio_test, factor_eval):
      batch_list = fct_prepare_batch(patches['training"])

      batch_list_test = batch_list[:ratio_test]
      batch_list_train = batch_list[ratio_test:]

      batch_cur_test = fct_extraction_batch(batch_list_test)
      batch_feature_test = fct_extraction_feature(batch_cur_test)

     for param in hyperparam_dict:
          for n_epoch in range(nb_epoch):
                for i, batch in enumerate(batch_list_train):
                     batch_cur_train = fct_extraction_batch(batch)
                     batch_feature_train = fct_extraction_feature(batch_cur_train) 
 
                     model.train(batch_feature_train)
                     
                    if i % factor_eval:
                          prediction = model.predict(batch_feature_test)
                          score_report = fct_evaluation(prediction, ground_truth)
                          save(score_report)
                          model.save(name)
```

**Comment by charleygros** [comment]
To use 'hyperopt' library for hyperparameter optimization, run:
```shell
pip install hyperopt
pip install pymongo
```

**Comment by charleygros** [comment]
Hyper parameters dict convention as Trainer Input:
- list of 2 float or int:
```python
'param_name': [low, high] --> hp.uniform('param_name', low, high)
```
- tuple of choices:
```python
'param_name':(choice_1, choice_2, ...) --> hp.choice('param_name', [choice_1, choice_2, ...])
```

(Related to: https://github.com/hyperopt/hyperopt/wiki/FMin)

**Comment by charleygros** [comment]
[Hyperparameter optimization] How to get results:

```python
with open(results_path + filename) as outfile:    
                trial = pickle.load(outfile)
                outfile.close()
            loss_list = [trial[i]['result']['loss'] for i in range(len(trial))]
            eval_time_list = [trial[i]['result']['eval_time'] for i in range(len(trial))]
```

**Comment by charleygros** [comment]
sct_detect_spinalcord USAGE:

```python
USAGE
sct_detect_spinalcord -i <folder> -d <file> -ofolder <folder> -n <int> -imodel <str> -threshold <float>

MANDATORY ARGUMENTS
 -i <folder>                  Folder containing testing images and their segmentation
 -d <file>                    Dictionary containing fname of testing images
 -ofolder <folder>            Path output                   
 -n <int>                     Number of images to process. Set '-n -1' to process all available 
                              images
 -imodel <str>                Model path                    
 -threshold <float>           Threshold value               

OPTIONAL ARGUMENTS
 -c {t1,t2}                   type of image contrast, t2: cord dark / CSF bright ; t1: cord bright 
                              / CSF dark.
 -v <int>                     Verbose Default value = 0     

EXAMPLE
sct_detect_spinalcord -i /Volumes/data_processing/bdeleener/machine_learning/large_nobrain_nopad/ -d /Volumes/data_processing/bdeleener/machine_learning/filemanager_large_nobrain_nopad/datasets.pbz2 -c t1 -ofolder /Users/chgroc/data/spine_detection/results3D/ -n 3 -imodel /Users/chgroc/data/spine_detection/results2D/model_t2_linear_000/LinearSVM_train.pkl -threshold 0.5 -v 0
```

Hard coded parameters:
- patch size
- initial_resolution
- initial_list_offset
- offset
- max_iter

**Comment by charleygros** [comment]
**Training Hyperparameters**
ISMRM 17 abstract - Update 11/09/2016

- msct_machine_learning_svm.py
```python
linear_svm_model = {'model_name': 'LinearSVM', 'model': Classifier_svm(svm.SVC),
                    'model_hyperparam':{'C': [1, 1000],
                                        'kernel': 'linear',
                                        'probability': True,
                                        'class_weight': ('balanced')}}

param_training = {'data_path_local': '/Volumes/data_processing/bdeleener/machine_learning/large_nobrain_nopad/',
                    'number_of_epochs': 1, 'patch_size': [32, 32],
                    'minibatch_size_train': None, # number for CNN, None for SVM
                    'minibatch_size_test': 1000,
                    'hyperopt': {'algo':tpe.suggest,        # Grid Search algorithm
                                'nb_eval':10,               # Nb max of param test
                                'fct': recall_score,       # Objective function
                                'eval_factor': 1,           # Evaluation rate
                                'ratio_dataset_eval':0.15,   # Ratio of training dataset dedicated to hyperParam validation
                                'ratio_img_eval':1.0,       # Ratio of patch per validation image
                                'ratio_img_train':0.3}}      # Ratio of patch per training image

coord_prepared_train, label_prepared_train = my_trainer.prepare_patches(my_trainer.fname_training_raw_images, 0.01)
coord_prepared_test, label_prepared_test = my_trainer.prepare_patches(my_trainer.fname_testing_raw_images, 1.0)
```

- train_patch_based_cnn.py
```python
params_cnn = {'patch_size': [32, 32],
              'number_of_channels': 1,
              'batch_size': 128,
              'number_of_features': [[32, 32], [64, 64]],
              'loss': 'categorical_crossentropy'
             }

cnn_model = {'model_name': 'CNN', 'model': KerasConvNet(params_cnn),
             'model_hyperparam': {'class_weight': ('balanced')}}

param_training = {'data_path_local': dataset_path,
                  'number_of_epochs': 10, 'patch_size': [32, 32], 'ratio_patch_per_img': 1.0,
                  'minibatch_size_train': 256, 'minibatch_size_test': 256,  # number for CNN, None for SVM
                  'hyperopt': {'algo': tpe.suggest,  # Grid Search algorithm
                               'nb_eval': 10,  # Nb max of param test
                               'fct': roc_auc_score,  # Objective function
                               'eval_factor': 10000,  # Evaluation rate
                               'ratio_dataset_eval': 0.25,
                               # Ratio of training dataset dedicated to hyperParam validation
                               'ratio_img_eval': 0.25,  # Ratio of patch per validation image
                               'ratio_img_train': 1.0}}

coord_prepared_train, label_prepared_train = my_trainer.prepare_patches(my_trainer.fname_training_raw_images, 1.0)
coord_prepared_test, label_prepared_test = my_trainer.prepare_patches(my_trainer.fname_testing_raw_images, 1.0)
```

**Comment by charleygros** [comment]
**3D SC detection**
ISMRM 17 abstract - Update 11/09/2016

- sct_detect_spinalcord_cnn.py
```python
grid_search_dict_t2={'patch_size':32,
                    'initial_resolution': [3, 3, 10],
                    'initial_resize': [0.1, 0.2],
                    'initial_list_offset': [[xv, yv, zv] for xv in range(-1,1) for yv in range(-1,1) for zv in range(-2,2) if [xv, yv, zv] != [0, 0, 0]],
                    'offset': [2,2,5],
                    'hyperparam_remove': [0.2, 0.01],
                    'hyperparam_search': [1, 3],
                    'hyperparam_highway': [0.2, 0.25],
                    'max_iter': 3
                    }


grid_search_dict_t2s={'patch_size':32,
                    'initial_resolution': [3, 3, 1],
                    'initial_resize': [0.4, 0.4],
                    'initial_list_offset': [[xv, yv, zv] for xv in range(-5,5) for yv in range(-5,5) 
                                                            for zv in range(-7,7) if [xv, yv, zv] != [0, 0, 0]],
                    'offset': [2,2,5],
                    'hyperparam_remove': [0.5, 0.007],
                    'hyperparam_search': [0, 2],
                    'hyperparam_highway': [0.3, 0.5],
                    'max_iter': 10
                    }
```

- sct_detect_spinalcord_svm.py
```python
grid_search_dict_t2={'patch_size':32,
                    'initial_resolution': [3, 3, 10],
                    'initial_resize': [0.1, 0.2],
                    'initial_list_offset': [[xv, yv, zv] for xv in range(-1,1) for yv in range(-1,1) for zv in range(-2,2) if [xv, yv, zv] != [0, 0, 0]],
                    'offset': [2,2,5],
                    'hyperparam_remove': [0.2, 0.01],
                    'hyperparam_search': [1, 2],
                    'hyperparam_highway': [0.2, 0.25],
                    'max_iter': 3
                    }

grid_search_dict_t2s={'patch_size':32,
                    'initial_resolution': [3, 3, 1],
                    'initial_resize': [0.4, 0.4],
                    'initial_list_offset': [[xv, yv, zv] for xv in range(-5,5) for yv in range(-5,5) 
                                                            for zv in range(-7,7) if [xv, yv, zv] != [0, 0, 0]],
                    'offset': [2,2,5],
                    'hyperparam_remove': [0.8, 0.008],
                    'hyperparam_search': [0, 2],
                    'hyperparam_highway': [0.3, 0.5],
                    'max_iter': 10
                    }
```

**Comment by joshuacwnewton** [comment]
Correct me if I'm wrong, but this issue appears to contain some preliminary work that would eventually lead to the CNN-based `sct_deepseg_sc` function.

Given that `sct_deepseg_sc` has been implemented, and their are more recent DL-based approaches in the works for spinal cord segmentation (e.g. SoftSeg), is it fair to say that this work is not relevant to modern-day SCT?

I'm going to close this issue for now, but if the work here is still relevant, please let me know.
